### PR TITLE
scaleway: provider

### PR DESCRIPTION
**Pull request by anubhavmishra**
* Adding scaleway provider

### CODE REVIEW

**Comment by anubhavmishra** [comment]
@magiconair Please do let me know if there is anything I missed :)

File: provider/scaleway/scaleway_discover.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	organization := args["organization"]
+	tagName := args["tag_name"]
+	apiKey := args["api_key"]
+	region := args["region"]
+
+	l.Printf("[INFO] discover-scaleway: Organization is %q", organization)
+	l.Printf("[INFO] discover-scaleway: Region is %q", region)
+
+	// Create a new API client
+	api, err := api.New(
```
Comment (magiconair) [inline provider/scaleway/scaleway_discover.go]: pls format as one line

File: provider/scaleway/scaleway_discover.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+		return nil, fmt.Errorf("discover-scaleway: %s", err)
+	}
+
+	// Get a list of private ips that match the tag name
+	return filterServersForTagName(servers, tagName), nil
+}
+
+func filterServersForTagName(servers *[]api.ScalewayServer, tagName string) []string {
+	var serverAddrs []string
+	for _, server := range *servers {
```
Comment (magiconair) [inline provider/scaleway/scaleway_discover.go]: can `servers` be `nil`?
pls add a debug log line for the servers you've found, e.g. like https://github.com/hashicorp/go-discover/blob/master/provider/aws/aws_discover.go#L120

File: provider/scaleway/scaleway_discover.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("discover-scaleway: %s", err)
+	}
+
+	// Get a list of private ips that match the tag name
+	return filterServersForTagName(servers, tagName), nil
+}
+
+func filterServersForTagName(servers *[]api.ScalewayServer, tagName string) []string {
+	var serverAddrs []string
```
Comment (magiconair) [inline provider/scaleway/scaleway_discover.go]: s/serverAddrs/addrs/

File: provider/scaleway/scaleway_discover.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+		apiKey,
+		region,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("discover-scaleway: %s", err)
+	}
+
+	// Currently fetching all servers since the API doesn't support
+	// filter options
+	servers, err := api.GetServers(true, 0)
```
Comment (magiconair) [inline provider/scaleway/scaleway_discover.go]: `0` is a magic number. What does it mean? same for `true`.

File: provider/scaleway/scaleway_discover.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	return filterServersForTagName(servers, tagName), nil
+}
+
+func filterServersForTagName(servers *[]api.ScalewayServer, tagName string) []string {
+	var serverAddrs []string
+	for _, server := range *servers {
+		if stringInSlice(tagName, server.Tags) {
+			l.Printf("[INFO] discover-scaleway: Found server (%d) - %s with private IP: %s",
+				server.Name, server.Hostname, server.PrivateIP)
+			serverAddrs = append(serverAddrs, server.PrivateIP)
```
Comment (magiconair) [inline provider/scaleway/scaleway_discover.go]: pls add a debug log line for each server and it's ip you've added, e.g. https://github.com/hashicorp/go-discover/blob/master/provider/aws/aws_discover.go#L145

File: discover.go [hunk 6]
```diff
@@ -38,6 +39,7 @@ var Providers = map[string]Provider{
 	"gce":          &gce.Provider{},
 	"os":           &os.Provider{},
 	"softlayer":    &softlayer.Provider{},
+	"scaleway":     &scaleway.Provider{},
```
Comment (magiconair) [inline discover.go]: pls keep alphabetical order

File: provider/scaleway/scaleway_discover.go [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+		return nil, fmt.Errorf("discover-scaleway: %s", err)
+	}
+
+	// Get a list of private ips that match the tag name
+	return filterServersForTagName(servers, tagName, l), nil
+}
+
+func filterServersForTagName(servers *[]api.ScalewayServer, tagName string, l *log.Logger) []string {
+	var serverAddrs []string
+	for _, server := range *servers {
```
Comment (magiconair) [inline provider/scaleway/scaleway_discover.go]: Can `servers` be `nil`?

File: provider/scaleway/scaleway_discover.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	if err != nil {
+		return nil, fmt.Errorf("discover-scaleway: %s", err)
+	}
+
+	// Get a list of private ips that match the tag name
+	return filterServersForTagName(servers, tagName, l), nil
+}
+
+func filterServersForTagName(servers *[]api.ScalewayServer, tagName string, l *log.Logger) []string {
+	var serverAddrs []string
```
Comment (magiconair) [inline provider/scaleway/scaleway_discover.go]: s/serverAddrs/addrs/

File: provider/scaleway/scaleway_discover.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+	// Currently fetching all servers since the API doesn't support
+	// filter options
+	servers, err := api.GetServers(true, 0)
+	if err != nil {
+		return nil, fmt.Errorf("discover-scaleway: %s", err)
+	}
+
+	// Get a list of private ips that match the tag name
+	return filterServersForTagName(servers, tagName, l), nil
```
Comment (magiconair) [inline provider/scaleway/scaleway_discover.go]: add DEBUG log line like https://github.com/hashicorp/go-discover/blob/master/provider/aws/aws_discover.go#L161

File: README.md [hunk 10]
```diff
@@ -28,6 +28,7 @@ function.
  * Microsoft Azure [Config options](https://github.com/hashicorp/go-discover/blob/master/provider/azure/azure_discover.go#L16-L30)
  * Openstack [Config options](https://github.com/hashicorp/go-discover/blob/master/provider/os/os_discover.go#L23-L38)
  * SoftLayer [Config options](https://github.com/hashicorp/go-discover/blob/master/provider/softlayer/softlayer_discover.go#L16-L25)
+ * Scaleway [Config options](https://github.com/hashicorp/go-discover/blob/master/provider/scaleway/scaleway_discover.go#L14-L22)
```
Comment (magiconair) [inline README.md]: pls keep alphabetical order

File: README.md [hunk 11]
```diff
@@ -52,6 +53,9 @@ provider=os tag_key=consul tag_value=server username=... password=... auth_url=.
 
 # SoftLayer
 provider=softlayer datacenter=dal06 tag_value=consul username=... api_key=...
+
```
Comment (magiconair) [inline README.md]: same here

**Comment by anubhavmishra** [comment]
@magiconair thank you will have the changes pushed soon!

File: provider/scaleway/scaleway_discover.go (hunk 9 above)
Comment (anubhavmishra) [reply provider/scaleway/scaleway_discover.go]: Added at the end of `filterServersForTagName` https://github.com/anubhavmishra/go-discover/blob/e742c50534c126f9455e1f5b8e8a96274d17f610/provider/scaleway/scaleway_discover.go#L74

**Comment by anubhavmishra** [comment]
@magiconair I will submit the vendored PR after this is merged.

File: provider/scaleway/scaleway_discover.go (hunk 4 above)
Comment (anubhavmishra) [reply provider/scaleway/scaleway_discover.go]: I wrote a detailed comment to explain how the `api.GetServers` works. Setting `0` lists all servers and the function won't panic 😄

File: provider/scaleway/scaleway_discover.go [hunk 12]
```diff
@@ -61,16 +61,18 @@ func (p *Provider) Addrs(args map[string]string, l *log.Logger) ([]string, error
 }
 
 func filterServersForTagName(servers *[]api.ScalewayServer, tagName string, l *log.Logger) []string {
-	var serverAddrs []string
-	for _, server := range *servers {
-		if stringInSlice(tagName, server.Tags) {
-			l.Printf("[INFO] discover-scaleway: Found server (%s) - %s with private IP: %s",
-				server.Name, server.Hostname, server.PrivateIP)
-			serverAddrs = append(serverAddrs, server.PrivateIP)
+	var addrs []string
```
Comment (magiconair) [inline provider/scaleway/scaleway_discover.go]:
```go
if len(servers) == 0 {
     return []string{}
}
var addrs []string
...
```

File: provider/scaleway/scaleway_discover.go [hunk 13]
```diff
@@ -61,16 +61,18 @@ func (p *Provider) Addrs(args map[string]string, l *log.Logger) ([]string, error
...
+			if stringInSlice(tagName, server.Tags) {
+				l.Printf("[DEBUG] discover-scaleway: Found server (%s) - %s with private IP: %s",
+					server.Name, server.Hostname, server.PrivateIP)
+				addrs = append(addrs, server.PrivateIP)
+			}
 		}
 	}
-
-	return serverAddrs
+	l.Printf("[DEBUG] discover-scaleway: Found ip addresses: %v", addrs)
```
Comment (magiconair) [inline provider/scaleway/scaleway_discover.go]: Please move this comment into the `Discover` function just before you return results.

**Comment by magiconair** [comment]
Thanks for submitting this!

**Comment by magiconair** [comment]
Merged manually in c98e36ab72ce62b7d8fbc7f7e76f9a60e163cb45

**Comment by anubhavmishra** [comment]
@magiconair you are so fast! thanks for adding the vendored libraries :) Sorry it took me longer haha

**Pull request closed without merging**